import os
import glob
import pandas as pd
from snakemake.utils import min_version

min_version('7.16.0')

configfile: '../config/hpcnode.yaml'

# Haplotype names. Used for finding haplotype fasta
HAPS = ['occ1', 'occ2', 'pall1', 'pall2']

# Define absolute path to pipeline resources
RESOURCES = os.path.abspath(config['resources'])

# Input files
HAPLOTYPE_FASTA_DIR = f"{RESOURCES}/haplotype_fastas/dovetail_v3"
HCN_LOCI_FASTA = f"{RESOURCES}/hcn_loci/T-repens_HCN_genes.fasta"
GENMAP_RESOURCE_DIR = f"{RESOURCES}/genMap"
REPBASE = f"{RESOURCES}/RepBaseRepeatMaskerEdition-20181026.tar.gz"
ORGANELLE_SEQS = f"{RESOURCES}/organelles/organelle_seqs.fast"
TRR_FIVE_SCAFFS = f"{RESOURCES}/TrR_v5_scaff_byLG.csv"
TRR_FIVE_FASTA = config['TrR_five_fasta']

RNASEQ_ACCESSIONS = pd.read_csv(f"{RESOURCES}/Trepens_rnaseq_accessions.txt", sep = '\t')['accession'].tolist()
KOOYERS_RNASEQ = [f for f in os.listdir(config['kooyers_rnaseq']) if f.startswith('TR')]

# Variables for specific rules
ALL_RNASEQ_SAMPLES = RNASEQ_ACCESSIONS + KOOYERS_RNASEQ

# Paths for output files
LOG_DIR = config['log_prefix']
BLAST_DIR = f"{config['results_prefix']}/blast"
SYNTENY_DIR = f"{config['results_prefix']}/synteny"
ANNOTATION_DIR = f"{config['results_prefix']}/annotation"
PROGRAM_RESOURCE_DIR = f"{config['results_prefix']}/program_resources"
REFERENCE_ASSEMBLIES_DIR = f"{config['results_prefix']}/reference_assemblies"

EXAMPLE_FASTA = HAPLOTYPE_FASTA_DIR + '/occ1/review_final_assembly.fasta' 

wildcard_constraints:
    acc = '|'.join(ALL_RNASEQ_SAMPLES)

include: 'rules/common.smk'

rule all:
    input:
        f'{BLAST_DIR}/blast.done',
        f'{SYNTENY_DIR}/synteny.done',
        f'{REFERENCE_ASSEMBLIES_DIR}/reference_assemblies.done',
        f"{ANNOTATION_DIR}/annotation.done",

include: 'rules/blast.smk'
include: 'rules/synteny.smk'
include: 'rules/assemble_references.smk'
include: 'rules/annotation.smk'

