# Dovetail Clover Genome (DCG)

TODO: Link manuscript/preprint when available
TODO: Figure out where raw haplotype FASTAs will be hosted upon publishing (NCBI?)
TODO: Link to NCBI Bioprojects
## Manuscript: Chromosome-level, haplotype-resolved assembly of white clover (*Trifolium repens*)
### Authors: James S. Santangelo, Paul Battlay, Brandon Hendrickson, Wen-Hsi Kuo, Kenneth M. Olsen, Nicholas J. Kooyers, Johnson, M.T. Johnson, Kathryn A. Hodgins, Rob W. Ness



## Description

This repository contains the code used for the assembly and annotation of a chromosome-level, haplotype-resolved assembly of white clover (*Trifolium repens*). The pipeline starts with input of the raw haplotype FASTAs sequenced and assembled by Dovetail Inc. using PacBio HiFi and Dovetail Omni-C technologies. It outputs two assemblies: (1) An unannotated, haplotype-resolved 'diploid' assembly (note: white clover is an allotetraploid, but exhibits disomic inheritance). This assembly is provided as separate Haplotype 1 and Haplotype 2 sequence files (see NCBI BioProjects XXX and YYY). (2) A functionally annotated, haploid mapping assembly generated by taking the longest chromosome from either haplptype 1 or 2 for each linkage ()group see NCBI BioProject ZZZ).

In addition to the raw dovetail haplotype FASTAs, the following dependencies/resources need to be obtained prior to running the pipeline.

1. `GeneMark-ES/ET/EP+ ver 4.71_lic` needs to be obtained from [here](http://topaz.gatech.edu/Genemark/license_download.cgi). `GeneMark` is licensed software and could not be included as part of the pipeline. Once downloaded, install `GeneMark` as recommended in their README file. Once installed, the following needs to be done:
    
    1. Include the `.gm_key` in the [workflow](./workflow) directory containing the main Snakefile used for running the pipeline
    2. Point to the `GeneMark` and `ProtHint` (included with `GeneMark`) programs in the [config](./config/hpcnode.yaml) file

2.  


